teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt

```diff
@@ -77,41 +131,95 @@ interface BuildLocator {
     fun pinnedOnly(): BuildLocator
 
     fun limitResults(count: Int): BuildLocator
-    
+
     fun sinceDate(date: Date) : BuildLocator
 
     fun latest(): Build?
+    fun all(): Sequence<Build>
+
+    @Deprecated(message = "use all() which returns lazy sequence",
+                replaceWith = ReplaceWith("all().toList()"))
     fun list(): List<Build>
+    @Deprecated(message = "use includeFailed()",
+                replaceWith = ReplaceWith("includeFailed()"))
+    fun withAnyStatus(): BuildLocator
 }
 
-data class ProjectId(val stringId: String)
+data class ProjectId(val stringId: String) {
+    override fun toString(): String = stringId
+}
 
-data class BuildId(val stringId: String)
+data class BuildId(val stringId: String) {
+    override fun toString(): String = stringId
+}
 
-data class QueuedBuildId(val stringId: String)
+data class TestId(val stringId: String) {
+    override fun toString(): String = stringId
+}
 
-data class ChangeId(val stringId: String)
+data class ChangeId(val stringId: String) {
+    override fun toString(): String = stringId
+}
 
-data class BuildConfigurationId(val stringId: String)
+data class BuildConfigurationId(val stringId: String) {
+    override fun toString(): String = stringId
+}
 
-data class VcsRootId(val stringId: String)
+data class VcsRootId(val stringId: String) {
+    override fun toString(): String = stringId
+}
+
+data class BuildProblemId(val stringId: String) {
+    override fun toString(): String = stringId
+}
+
+data class BuildProblemType(val stringType: String) {
+    override fun toString(): String = stringType
+
+    val isSnapshotDependencyError: Boolean
+        get() = stringType == "SNAPSHOT_DEPENDENCY_ERROR_BUILD_PROCEEDS_TYPE" ||
+                stringType == "SNAPSHOT_DEPENDENCY_ERROR"
+
+    companion object {
+        val FAILED_TESTS = BuildProblemType("TC_FAILED_TESTS")
+    }
+}
 
 interface Project {
     val id: ProjectId
     val name: String
     val archived: Boolean
-    val parentProjectId: ProjectId
+    val parentProjectId: ProjectId?
 
     /**
      * Web UI URL for user, especially useful for error and log messages
      */
+    fun getHomeUrl(branch: String? = null): String
+    fun getTestHomeUrl(testId: TestId): String
+
+    val childProjects: List<Project>
+    val buildConfigurations: List<BuildConfiguration>
+    val parameters: List<Parameter>
+
+    fun setParameter(name: String, value: String)
+
+    fun createVcsRoot(id: VcsRootId, name: String, type: VcsRootType, properties: Map<String, String>): VcsRoot
+    fun createProject(id: ProjectId, name: String): Project
+    fun createBuildConfiguration(buildConfigurationDescriptionXml: String): BuildConfiguration
```

Comment:
We need to give some information about format of `buildConfigurationDescriptionXml`, at least link to some documentation. Currently it's absolutely obscure.

Reply:
I'll add a javadoc how to get it, but not much more.